src/alpaka/Framework/WaitingTaskHolder.h

```diff
@@ -72,20 +95,29 @@ namespace edm {
       if (iExcept) {
         m_task->dependentTaskFailed(iExcept);
       }
-      //spawn can run the task before we finish
+      //task_group::run can run the task before we finish
       // doneWaiting and some other thread might
       // try to reuse this object. Resetting
       // before spawn avoids problems
       auto task = m_task;
       m_task = nullptr;
       if (0 == task->decrement_ref_count()) {
-        tbb::task::spawn(*task);
+        m_group->run([task]() {
+          TaskSentry s{task};
+          task->execute();
+        });
```

Comment:
Trying to understand what this pattern does:
  - `m_group` is a pointer to the `tbb::task_group` that will run the task
  - `m_task` is the a pointer to the `WaitingTask` to be run
  - `task` is a copy of `m_task` (the comment above explains why the extra copy is needed)
  - `m_group` takes a functor (here a lambda) to be run
  - the lambda makes a copy of `task` 
  - the `TaskSentry` takes ownership of the `task`, guaranteeing that it will be deleted after it has run (also in case an exception is thrown)
  - the task is executed inside the lambda inside the `task_group`
  
The only confusing aspect is that `TaskSentry` will delete the task once the lambda is complete, except if it is a `FinalWaitingTask`, in which case the destructor of the `TaskSentry` effectively does not do anything.

Reply:
That difference comes from objects of all other task types than `FinalWaitingTask` being allocated with `new` and intended to be deleted right after the task execution has finished. The `FinalWaitingTask` object is allocated on the stack of the function that initiates the "asynchronous region" and does a wait on the `tbb::task_group`. It must be kept alive until all the asynchronous tasks of the region have finished (e.g. to check if exception was thrown).